gk/fib.c

```diff
@@ -60,31 +65,90 @@ gk_lpm_add_ipv6_route(uint8_t *ip,
  * callers must ensure that the entry is not being used.
  */
 static inline int
-initialize_ether_cache(struct ether_cache *eth_cache)
+clear_ether_cache(struct ether_cache *eth_cache)
 {
-	memset(eth_cache, 0, sizeof(*eth_cache));
+	memset(eth_cache->fields_to_clear, 0, sizeof(*eth_cache) -
+		offsetof(struct ether_cache, fields_to_clear));
 
 	return 0;
 }
 
-/*
- * Fill up the Ethernet cached header.
- * Note that the destination MAC address should be filled up by LLS.
- */
-static inline void
-fill_up_ether_cache_locked(struct ether_cache *eth_cache,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	if (pcall_again == NULL) {
+		clear_ether_cache(eth_cache);
+		return;
```

Comment:
We must make field `ref_cnt` of `eth_cache` atomic. The reason is very subtle. Consider that `ether_cache_put()` is called on an Ethernet cache entry whose field `ref_cnt` is `1`. `ether_cache_put()` eventually calls `put_arp()` (or `put_nd` for IPv6). Given that `put_arp()` sends a message to the LLS block, and the LLS block may be busy with other events, it can take a while for `gk_arp_and_nd_req_cb()` to run for the last time in order to free the ethernet cache entry. Thus, the LLS block may run `gk_arp_and_nd_req_cb()` at the same time as a GK block runs `get_new_ether_cache_locked()`. `get_new_ether_cache_locked()` looks for entries that a free by checking that `ref_cnt == 0`. Since `clear_ether_cache()` in `gk_arp_and_nd_req_cb()` sets the memory without a synchronization mechanism, we are vulnerable to likely rare race condition that depends on specifics of the hardware.

My proposed solution to avoid depending on the hardware is (1) to make field `ref_cnt` of `eth_cache` atomic, (2) move field `ref_cnt` just before field `fields_to_clear`, (3) make `clear_ether_cache()` set field `ref_cnt` to zero just before returning. Of course, once field `ref_cnt` of `eth_cache` becomes atomic, one has to adjust the code that reads or writes this field to deal with the fact that it's now atomic.

`clear_ether_cache()` should log a serious message before setting field `ref_cnt` to zero if `ref_cnt` is not `1`.

Finally, we should have a comment on the definition of field `ref_cnt` in `eth_cache` explaining the reason it is atomic is that it's handled by the LLS block and the GK blocks. So we don't forget this intricate relationship.